google/cloud/internal/async_read_write_stream_auth.h

```diff
@@ -28,79 +29,97 @@ namespace internal {
 
 template <typename Request, typename Response>
 class AsyncStreamingReadWriteRpcAuth
-    : public AsyncStreamingReadWriteRpc<Request, Response>,
-      public std::enable_shared_from_this<
-          AsyncStreamingReadWriteRpcAuth<Request, Response>> {
+    : public AsyncStreamingReadWriteRpc<Request, Response> {
  public:
   using StreamFactory = std::function<
-      std::shared_ptr<AsyncStreamingReadWriteRpc<Request, Response>>(
+      std::unique_ptr<AsyncStreamingReadWriteRpc<Request, Response>>(
           std::unique_ptr<grpc::ClientContext>)>;
 
   AsyncStreamingReadWriteRpcAuth(
       std::unique_ptr<grpc::ClientContext> context,
       std::shared_ptr<GrpcAuthenticationStrategy> auth, StreamFactory factory)
-      : context_(std::move(context)),
-        auth_(std::move(auth)),
-        factory_(std::move(factory)) {}
+      : auth_(std::move(auth)),
+        state_(std::make_shared<SharedState>(std::move(factory), std::move(context))) {}
 
   void Cancel() override {
-    if (context_) return context_->TryCancel();
-    if (stream_) return stream_->Cancel();
+    state_->Cancel();
   }
 
   future<bool> Start() override {
     using Result = StatusOr<std::unique_ptr<grpc::ClientContext>>;
 
-    auto weak =
-        std::weak_ptr<AsyncStreamingReadWriteRpcAuth>(this->shared_from_this());
-    return auth_->AsyncConfigureContext(std::move(context_))
+    std::unique_ptr<grpc::ClientContext> context;
+    {
+      std::lock_guard<std::mutex> g{state_->mu};
```

Comment:
I think I would have written:

```cc
std::unique_lock<std::mutex> g{state_->mu}
auto context = std::move(state_->initial_context);
g.unlock();
auto weak = ...
...
```

Alternatively, add a `ReleaseContext()` function to `SharedState`?

Reply:
Personal preference to avoid unlocking methods because scopes more clearly deliniate the locked region. But a release method is cleaner